plugins/bnr-dp/fu-bnr-dp-device.c

```diff
@@ -0,0 +1,748 @@
+			    fu_struct_bnr_dp_aux_rx_header_get_checksum(header),
+			    actual_checksum);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* read only 2 bytes from the header offset to receive the status */
+static FuStructBnrDpAuxStatus *
+fu_bnr_dp_device_read_status(FuBnrDpDevice *self, GError **error)
+{
+	guint8 buf[FU_STRUCT_BNR_DP_AUX_STATUS_SIZE] = {0};
+	g_autoptr(FuStructBnrDpAuxStatus) status = NULL;
+
+	/* only read the first 2 bytes of the header to check status bits. */
+	if (!fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+				  FU_BNR_DP_DEVICE_HEADER_OFFSET,
+				  buf,
+				  sizeof(buf),
+				  FU_BNR_DP_DEVICE_DPAUX_TIMEOUT_MSEC,
+				  error))
+		return NULL;
+
+	status = fu_struct_bnr_dp_aux_status_parse(buf, sizeof(buf), 0, error);
+	if (status == NULL)
+		return NULL;
+
+	return g_steal_pointer(&status);
+}
+
+static gboolean
+fu_bnr_dp_device_poll_status(FuBnrDpDevice *self, GError **error)
+{
+	for (gint i = 0;; i++) {
+		g_autoptr(FuStructBnrDpAuxStatus) status = NULL;
+
+		status = fu_bnr_dp_device_read_status(self, error);
+		if (status == NULL)
+			return FALSE;
+
+		if (!fu_bnr_dp_device_eval_result(status, error))
+			return FALSE;
+		if (!fu_bnr_dp_device_is_busy(status))
+			return TRUE;
+
+		if (i > FU_BNR_DP_DEVICE_POLL_MAX_TOTAL) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_BUSY,
+					    "polling device reached timeout");
+			return FALSE;
+		}
+		if (i > FU_BNR_DP_DEVICE_POLL_MAX_FAST)
+			g_usleep(FU_BNR_DP_DEVICE_POLL_INTERVAL_USEC);
+	}
+}
+
+static GByteArray *
+fu_bnr_dp_device_exec_cmd(FuBnrDpDevice *self,
+			  FuBnrDpOpcodes opcode,
+			  FuBnrDpModuleNumber module_number,
+			  guint16 offset,
+			  GError **error)
+{
+	g_autoptr(GByteArray) r = g_byte_array_new();
+	g_autoptr(FuStructBnrDpAuxRequest) request = NULL;
+
+	request = fu_bnr_dp_device_build_request(opcode, module_number, offset, 0, error);
+	if (request == NULL)
+		return NULL;
+
+	if (!fu_bnr_dp_device_write_request(self, request, NULL, 0, error))
+		return NULL;
+
+	if (!fu_bnr_dp_device_poll_status(self, error)) {
+		g_prefix_error(error,
+			       "command %s to module %s at offset 0x%X: ",
+			       fu_bnr_dp_opcodes_to_string(opcode),
+			       fu_bnr_dp_module_number_to_string(module_number),
+			       offset);
+		return NULL;
+	}
+
+	if (!fu_bnr_dp_device_read_response(self, r, error))
+		return NULL;
+
+	return g_steal_pointer(&r);
+}
+
+static GByteArray *
+fu_bnr_dp_device_read_data(FuBnrDpDevice *self,
+			   FuBnrDpOpcodes opcode,
+			   FuBnrDpModuleNumber module_number,
+			   gsize offset,
+			   gsize size,
+			   FuProgress *progress,
+			   GError **error)
+{
+	g_autoptr(GByteArray) r = g_byte_array_sized_new(size);
+	const guint16 start = offset / FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE;
+	const guint16 end = (offset + size) / FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE;
+
+	g_assert_cmpint(offset % FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE, ==, 0);
+	g_assert_cmpint(size % FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE, ==, 0);
```

Comment:
these only work in self test code -- what you want is `g_return_val_if_fail` as g_assert() is also not allowed here.